python/cudf_polars/cudf_polars/experimental/rapidsmpf/shuffle.py

```diff
@@ -67,9 +61,63 @@ def _release_shuffle_id(op_id: int) -> None:
         _shuffle_id_vacancy.add(op_id)
 
 
-class LocalShuffle:
+class ReserveOpIDs:
+    """
+    Context manager to reserve shuffle IDs for pipeline execution.
+
+    Parameters
+    ----------
+    ir : IR
+        The root IR node of the pipeline.
+
+    Notes
+    -----
+    This context manager:
+    1. Identifies all Shuffle nodes in the IR
+    2. Reserves shuffle IDs from the vacancy pool
+    3. Creates a mapping from IR nodes to their reserved IDs
+    4. Releases all IDs back to the pool on __exit__
+    """
+
+    def __init__(self, ir: IR):
+        # Collect all Shuffle nodes.
+        # NOTE: We will also need to collect Repartition,
+        # and Join nodes to support multi-GPU execution.
+        self.shuffle_nodes: list[IR] = [
```

Comment:
When we do support `Repartition`, will we store them in the same list (`shuffle_nodes`) / dictionary (`shuffle_id_map`) as the shuffle nodes? I'm wonder if we can easily future proof the interface at all, so that we don't need a ton of noisy changes when we go to support other types.

Reply:
Yeah, it's the same list. We can just call it `collective_nodes` to make that clearer.

Reply:
I renamed everything to `collective_id`/`collective_nodes` and refactored some of the common logic into a separate file.